File: Day319_2025/Q319_GFG.java
class Q319_GFG {
    public int minCutCost(int n, int[] cuts) {

        Arrays.sort(cuts);

        int m = cuts.length;
        int[] arr = new int[m + 2];

        arr[0] = 0;
        for (int i = 0; i < m; i++) {
            arr[i + 1] = cuts[i];
        }
        arr[m + 1] = n;

        int[][] dp = new int[m + 2][m + 2];

        for (int l = m; l >= 1; l--) {
            for (int r = 1; r <= m; r++) {

                if (l > r) {
                    dp[l][r] = 0;
                    continue;
                }

                int ans = Integer.MAX_VALUE;

                for (int i = l; i <= r; i++) {

                    int cost = arr[r + 1] - arr[l - 1];

                    int left = (l > i - 1) ? 0 : dp[l][i - 1];
                    int right = (i + 1 > r) ? 0 : dp[i + 1][r];

                    ans = Math.min(ans, cost + left + right);
                }

                dp[l][r] = ans;
            }
        }

        return dp[1][m];
    }
}

File: Day319_2025/Q319_GFG.py
class Solution:
    def minCutCost(self, n, cuts):
        cuts.sort()
        cuts = [0] + cuts + [n]

        m = len(cuts) - 2  # original number of cuts

        dp = [[0] * (m + 2) for _ in range(m + 2)]

        for l in range(m, 0, -1):
            for r in range(1, m + 1):

                if l > r:
                    dp[l][r] = 0
                    continue

                ans = float('inf')

                for i in range(l, r + 1):
                    cost = cuts[r + 1] - cuts[l - 1]

                    left = 0 if l > i - 1 else dp[l][i - 1]
                    right = 0 if i + 1 > r else dp[i + 1][r]

                    ans = min(ans, cost + left + right)

                dp[l][r] = ans

        return dp[1][m]
